plugin/rpc_server.go

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package plugin
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/apache/brooklyn-client/app"
+	"github.com/apache/brooklyn-client/command"
+	"github.com/apache/brooklyn-client/command_metadata"
+	"github.com/apache/brooklyn-client/command_runner"
+	"github.com/apache/brooklyn-client/scope"
+	"github.com/codegangsta/cli"
+	"net"
+	"net/rpc"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+type RpcServer struct {
+	listener    net.Listener
+	stopChannel chan struct{}
+	RpcCommand  *RpcCommand
+}
+
+type Factory interface {
+	GetByCmdName(cmdName string) (cmd command.Command, err error)
+	GetBySubCmdName(cmdName string, subCmdName string) (cmd command.Command, err error)
+	CommandMetadatas() []command_metadata.CommandMetadata
+}
+
+type Runner interface {
+	RunCmdByName(cmdName string, c *cli.Context) (err error)
+	RunSubCmdByName(cmdName string, subCommand string, c *cli.Context) (err error)
+}
+
+type RpcCommand struct {
+	PluginMetadata *PluginMetadata
+	commandFactory Factory
+}
+
+func NewRpcServer(commandFactory Factory) (*RpcServer, error) {
+	rpcServer := &RpcServer{
+		RpcCommand: &RpcCommand{
+			PluginMetadata: &PluginMetadata{},
+			commandFactory: commandFactory,
+		},
+	}
+	return rpcServer, nil
+}
+
+func (rpcServer *RpcServer) Start() error {
+	var err error
+
+	rpcServer.stopChannel = make(chan struct{})
+	rpc.Register(rpcServer.RpcCommand)
+	rpcServer.listener, err = net.Listen("tcp", "127.0.0.1:4567")
```

Comment:
Better to get an arbitrary free port rather than hard code this.  I believe if you use the value `:0` here then the os will assign you a free port of its own choosing.